pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriter.java

```diff
@@ -259,138 +353,65 @@ private void internalAsyncAddData(T data, AddDataCallback callback, Object ctx){
+        // If other thread already called "close()", so do nothing.
         if (!STATE_UPDATER.compareAndSet(this, State.OPEN, State.CLOSING)){
-            // Other thread also calling "close()".
             return;
         }
         // Cancel pending tasks and release resources.
         singleThreadExecutorForWrite.execute(() -> {
-            if (state == State.CLOSED){
-                return;
-            }
-            // Failure callback to pending request.
-            // If some request has been flushed, Bookie triggers the callback.
-            failureCallbackByContextAndRecycle(this.flushContext, BUFFERED_WRITER_CLOSED_EXCEPTION);
-            // Cancel task that schedule at fixed rate trig flush.
-            if (timeout == null){
-                log.error("Cancel timeout-task that schedule at fixed rate trig flush failure. The field-timeout"
-                        + " is null. managedLedger: " + managedLedger.getName());
-            } else if (timeout.isCancelled()){
-                // TODO How decisions the timer-task has been finished ?
-                this.state = State.CLOSED;
-            } else {
-                if (this.timeout.cancel()) {
-                    this.state = State.CLOSED;
+            try {
+                if (state == State.CLOSED) {
+                    return;
+                }
+                // If some requests are flushed, BK will trigger these callbacks, and the remaining requests in should
+                // fail.
+                failureCallbackByContextAndRecycle(flushContext, BUFFERED_WRITER_CLOSED_EXCEPTION);
+                // Cancel the timing task.
+                if (timeout == null) {
+                    log.error("Cancel timeout-task that schedule at fixed rate trig flush failure. The field-timeout"
+                            + " is null. managedLedger: " + managedLedger.getName());
+                } else if (timeout.isCancelled()) {
+                    // TODO How decisions the timer-task has been finished ?
+                    STATE_UPDATER.set(this, State.CLOSED);
                 } else {
-                    // Cancel task failure, The state will stay at CLOSING.
-                    log.error("Cancel timeout-task that schedule at fixed rate trig flush failure. The state will"
-                            + " stay at CLOSING. managedLedger: " + managedLedger.getName());
+                    if (this.timeout.cancel()) {
+                        STATE_UPDATER.set(this, State.CLOSED);
+                    } else {
+                        // Cancel task failure, The state will stay at CLOSING.
+                        log.error("Cancel timeout-task that schedule at fixed rate trig flush failure. The state will"
+                                + " stay at CLOSING. managedLedger: " + managedLedger.getName());
+                    }
                 }
+            } catch (Exception e){
+                log.error("Close Txn log buffered writer fail", e);
```

Comment:
Actually, I'm not sure it makes sense here - why would you not want to propagate it upwards?

Reply:
Good idea, already fixed